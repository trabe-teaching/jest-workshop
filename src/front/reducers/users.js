import { USERS_UPDATED } from "action/creators";

const initialState = {
  repo: {},
  lightSiders: [],
  darkSiders: [],
};

const lightSider = user => user.type === "light";
const darkSider = user => user.type === "dark";

const id = user => user.id;
const indexById = col => col.reduce((acc, e) => ({ ...acc, [id(e)]: e }), {});

export default (state = initialState, { type, payload }) => {
  switch (type) {
    case USERS_UPDATED:
      const updatedRepo = { ...state.repo, ...indexById(payload.users) };
      const users = Object.values(updatedRepo);

      return {
        repo: updatedRepo,
        lightSiders: users.filter(lightSider).map(id),
        darkSiders: users.filter(darkSider).map(id),
      };

    default:
      return state;
  }
};

const hydrate = (state, ids) =>
  ids.map(id => {
    const { actions, master, ...user } = state.repo[id];

    return {
      ...user,
      master: master && state.repo[master],
      canSeek: actions.includes("seek"),
      canTurn: actions.includes("turn"),
      canBetray: actions.includes("betray"),
      canBanish: actions.includes("banish"),
    };
  });

export const getLightSiders = state => hydrate(state, state.lightSiders);
export const getDarkSiders = state => hydrate(state, state.darkSiders);
