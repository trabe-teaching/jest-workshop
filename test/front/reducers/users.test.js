import deepFreeze from "deepfreeze";
import users, { getLightSiders, getDarkSiders } from "reducers/users";

const state = {
  repo: {
    1: {
      id: 1,
      name: "Luke Skywalker",
      type: "light",
      rank: "master",
      status: "alive",
      email: "luke@ach.to",
      actions: ["banish"],
    },
    2: {
      id: 2,
      name: "Yoda",
      type: "light",
      rank: "master",
      status: "ghost",
      email: "yoda@dagobah.com",
      actions: [],
    },
    3: {
      id: 3,
      name: "Ben Solo",
      type: "light",
      rank: "apprentice",
      status: "alive",
      master: 1,
      email: "ben.solo@gmail.com",
      actions: ["turn", "betray", "seekMaster"],
    },
  },
  lightSiders: [1, 2],
  darkSiders: [3],
};

deepFreeze(state);

describe("front/reducers/application", () => {
  describe("reducer", () => {
    it("has a default state", () => {
      expect(users(undefined, { type: "@init" })).toMatchSnapshot();
    });

    it("reduces nothing", () => {
      expect(users(state, { type: "NOPE" })).toEqual(state);
    });

    it("reduces REQUEST", () => {
      const action = {
        type: "USERS_UPDATED",
        payload: {
          users: [
            {
              id: 1,
              name: "Luke Skywalker",
              type: "light",
              rank: "master",
              status: "ghost",
              email: "luke@ach.to",
            },
          ],
        },
      };

      expect(users(state, action)).toMatchSnapshot();
    });
  });

  describe("#getLightSiders", () => {
    it("returns ligh users", () => {
      expect(getLightSiders(state)).toMatchSnapshot();
    });

    it("returns dark users", () => {
      expect(getDarkSiders(state)).toMatchSnapshot();
    });
  });
});
