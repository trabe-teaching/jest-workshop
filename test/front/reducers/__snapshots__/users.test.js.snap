// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`front/reducers/application #getLightSiders returns dark users 1`] = `
Array [
  Object {
    "canBanish": false,
    "canBetray": true,
    "canSeek": false,
    "canTurn": true,
    "email": "ben.solo@gmail.com",
    "id": 3,
    "master": Object {
      "actions": Array [
        "banish",
      ],
      "email": "luke@ach.to",
      "id": 1,
      "name": "Luke Skywalker",
      "rank": "master",
      "status": "alive",
      "type": "light",
    },
    "name": "Ben Solo",
    "rank": "apprentice",
    "status": "alive",
    "type": "light",
  },
]
`;

exports[`front/reducers/application #getLightSiders returns ligh users 1`] = `
Array [
  Object {
    "canBanish": true,
    "canBetray": false,
    "canSeek": false,
    "canTurn": false,
    "email": "luke@ach.to",
    "id": 1,
    "master": undefined,
    "name": "Luke Skywalker",
    "rank": "master",
    "status": "alive",
    "type": "light",
  },
  Object {
    "canBanish": false,
    "canBetray": false,
    "canSeek": false,
    "canTurn": false,
    "email": "yoda@dagobah.com",
    "id": 2,
    "master": undefined,
    "name": "Yoda",
    "rank": "master",
    "status": "ghost",
    "type": "light",
  },
]
`;

exports[`front/reducers/application reducer has a default state 1`] = `
Object {
  "darkSiders": Array [],
  "lightSiders": Array [],
  "repo": Object {},
}
`;

exports[`front/reducers/application reducer reduces REQUEST 1`] = `
Object {
  "darkSiders": Array [],
  "lightSiders": Array [
    1,
    2,
    3,
  ],
  "repo": Object {
    "1": Object {
      "email": "luke@ach.to",
      "id": 1,
      "name": "Luke Skywalker",
      "rank": "master",
      "status": "ghost",
      "type": "light",
    },
    "2": Object {
      "actions": Array [],
      "email": "yoda@dagobah.com",
      "id": 2,
      "name": "Yoda",
      "rank": "master",
      "status": "ghost",
      "type": "light",
    },
    "3": Object {
      "actions": Array [
        "turn",
        "betray",
        "seekMaster",
      ],
      "email": "ben.solo@gmail.com",
      "id": 3,
      "master": 1,
      "name": "Ben Solo",
      "rank": "apprentice",
      "status": "alive",
      "type": "light",
    },
  },
}
`;
